lib/modules/manager/bazel-module/rules.ts

```diff
@@ -0,0 +1,146 @@
+  depType: string;
+  depName: string;
+}
+
+export interface OverridePackageDep extends BasePackageDep {
+  // This value is set as the skipReason on the bazel_dep PackageDependency.
+  bazelDepSkipReason: SkipReason;
+}
+
+type BazelModulePackageDep = BasePackageDep | OverridePackageDep;
+
+function isOverride(value: BazelModulePackageDep): value is OverridePackageDep {
+  return (value as OverridePackageDep).bazelDepSkipReason !== undefined;
+}
+
+// This function exists to remove properties that are specific to
+// OverridePackageDep. In theory, there is no harm in leaving the properties
+// as it does not invalidate the PackageDependency interface. However, it might
+// be surprising to someone outside the bazel-module code to see the extra
+// properties.
+export function overrideToPackageDependency(
+  override: Partial<OverridePackageDep>
+): PackageDependency {
+  const copy: Partial<OverridePackageDep> = structuredClone(override);
+  delete copy.bazelDepSkipReason;
+  return copy;
+}
+
+const BazelDepToPackageDep = RecordFragmentSchema.extend({
+  children: z.object({
+    rule: StringFragmentSchema.extend({
+      value: z.literal('bazel_dep'),
+    }),
+    name: StringFragmentSchema,
+    version: StringFragmentSchema,
+  }),
+}).transform(
+  ({ children: { rule, name, version } }): BasePackageDep => ({
+    datasource: BazelDatasource.id,
+    depType: rule.value,
+    depName: name.value,
+    currentValue: version.value,
+  })
+);
+
+const GitOverrideToPackageDep = RecordFragmentSchema.extend({
+  children: z.object({
+    rule: StringFragmentSchema.extend({
+      value: z.literal('git_override'),
+    }),
+    module_name: StringFragmentSchema,
+    remote: StringFragmentSchema,
+    commit: StringFragmentSchema,
+  }),
+}).transform(
+  ({
+    children: { rule, module_name: moduleName, remote, commit },
+  }): OverridePackageDep => {
+    const override: OverridePackageDep = {
+      depType: rule.value,
+      depName: moduleName.value,
+      bazelDepSkipReason: 'git-dependency',
+      currentDigest: commit.value,
+    };
+    if (isGithubRemote(remote.value)) {
+      override.datasource = GithubTagsDatasource.id;
+      override.packageName = githubPackageName(remote.value);
+    } else {
+      override.skipReason = 'unsupported-datasource';
+    }
+    return override;
+  }
+);
+
+export const RuleToBazelModulePackageDep = z.union([
+  BazelDepToPackageDep,
+  GitOverrideToPackageDep,
+]);
+
+function isGithubRemote(remote: string): boolean {
+  return is.truthy(parseGithubUrl(remote));
+}
+
+const githubRemoteRegex = regEx(
+  /^https:\/\/github\.com\/(?<packageName>[^/]+\/.+)$/
+);
+function githubPackageName(remote: string): string | undefined {
+  return parseGithubUrl(remote)?.match(githubRemoteRegex)?.groups?.packageName;
+}
+
+function collectByModule(
+  packageDeps: BazelModulePackageDep[]
+): BazelModulePackageDep[][] {
+  const rulesByModule = packageDeps.reduce((map, pkgDep) => {
```

Comment:
use for-of loop

Reply:
OK. Can you explain why we should not use `.reduce()`?